### PR TITLE
Adding vector fitting and export of equivalent SPICE subcircuits

### PR DESCRIPTION
**Pull request by Vinc0110**
This PR adds the class VectorFitting, which implements a vector fitting algorithm to fit the frequency responses (S-matrix) in a `Network` to a sum of rational basis functions. Those rational basis functions can easily be translated into equivalent passive or active impedances or admittances, enabling the modelling of the original frequency response in any kind of circuit simulator, e.g. SPICE simulations. 

A routine is provided to automatically create such an equivalent SPICE subcircuit once the vector fit was performed.

A first draft of an example notebook is added and there is also a first and very simple testcase for the vector fitting.

**Edit**: 
- Example update: 2 examples included
- Tutorial is added

### CODE REVIEW

File: skrf/vectorFitting.py [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+        mplt.semilogy(np.arange(np.alen(self.delta_max_history)) + 1, self.delta_max_history, color='darkblue')
+        mplt.xlabel('Iteration step')
+        mplt.ylabel('Max. relative change', color='darkblue')
+        ax2 = mplt.twinx()
+        ax2.plot(np.arange(np.alen(self.d_res_history)) + 1, self.d_res_history, color='orangered')
+        ax2.set_ylabel('Residue', color='orangered')
+        mplt.tight_layout()
+        mplt.show()
+
+    def write_spice_subckt_s(self, file):
```
Comment (jhillairet) [inline skrf/vectorFitting.py:601]: Suggestion: `write_spice_subcircuit_s` unless `subckt` is a conventional name in SPICE (?). It only adds few letter and to me it's clearer

File: skrf/vectorFitting.py [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+        Transactions on Power Delivery, vol. 14, no. 3, pp. 1052-1061, July 1999
+    [2] B. Gustavsen, "Improving the Pole Relocating Properties of Vector Fitting", IEEE Transactions on Power Delivery,
+        vol. 21, no. 3, pp. 1587-1592, July 2006
+    [3] D. Deschrijver, M. Mrozowski, T. Dhaene, D. De Zutter, "Marcomodeling of Multiport Systems Using a Fast
+        Implementation of the Vector Fitting Method", IEEE Microwave and Wireless Components Letters, vol. 18, no. 6,
+        pp. 383-385, June 2008
+    [4] https://www.sintef.no/projectweb/vectorfitting/
+    """
+
+    def __init__(self, network):
```
Comment (jhillairet) [inline skrf/vectorFitting.py:23]: Would it be useful to be able to construct the `VectorFitting` object from given poles and zeros? (just a question, I have no idea of potential use cases)

File: TODO.rst [hunk 3]
```diff
@@ -37,21 +37,24 @@ Calibration
...
 
 vi 
 --------------------
 * re-write VNA's coherently 
 * use consistent method names so VNA's are inter-changable as much as possible
 * make a generic VNA that uses IDN? to load the correct class (durrants idea)
 
+VectorFitting
+-------------
+* implement passivity evaluation and enforcement routines
```
Comment (jhillairet) [inline TODO.rst:50]: You remind me that TODO file exists, thanks ;)
It is probably obsolete in few things.

**Review by jhillairet — COMMENTED** [review]
Thanks you for this very nice addition! 

I've made few suggestions concerning docstring and function names, in order to (or trying to...) make it coherent with the rest of the skrf API.

File: skrf/vectorFitting.py [hunk 4]
```diff
@@ -0,0 +1,765 @@
...
+        """
+        Creates a VectorFitting instance based on a supplied scikit-rf Network instance containing the frequency
+        responses of the N-port.
+
+        :param network:
+            scikit-rf Network instance of the N-port to be fitted.
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py:30]: Sure. I'll do that soon in another commit

File: skrf/vectorFitting.py [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        self.poles = None
+        self.zeros = None
+        self.proportional_coeff = None
+        self.constant_coeff = None
+        self.max_iterations = 100
+        self.max_tol = 1e-6
+        self.d_res_history = []
+        self.delta_max_history = []
+
+    def vectorfit(self, n_poles_real=2, n_poles_cmplx=2, init_pole_spacing='lin', parameter_type='S',
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: Done in 5981299be06c6f9ac70790e63e4699ba0363ada6

File: skrf/vectorFitting.py [hunk 6]
```diff
@@ -0,0 +1,765 @@
...
+            return
+
+        filename = self.network.name
+
+        print('Exporting results as pickled NumPy array to {}'.format(path))
+        np.savez(os.path.join(path, 'coefficients_{}'.format(filename)),
+                 poles=self.poles, zeros=self.zeros, proportionals=self.proportional_coeff,
+                 constants=self.constant_coeff)
+
+    def import_npz(self, file):
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: Done in 5981299be06c6f9ac70790e63e4699ba0363ada6

File: skrf/vectorFitting.py [hunk 7]
```diff
@@ -0,0 +1,765 @@
...
+
+        # save poles, zeros, d, e in actual frequencies (un-normalized)
+        self.poles = poles * norm
+        self.zeros = np.array(zeros) * norm
+        self.constant_coeff = np.array(constant_coeff)
+        self.proportional_coeff = np.array(proportional_coeff) / norm
+
+        print('\n### Vector fitting finished.\n')
+
+    def export_npz(self, path):
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: Done in 5981299be06c6f9ac70790e63e4699ba0363ada6

File: skrf/vectorFitting.py [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+        for i, pole in enumerate(self.poles):
+            if np.imag(pole) == 0.0:
+                # real pole
+                resp += zeros[i] / (s - pole)
+            else:
+                # complex conjugate pole
+                resp += zeros[i] / (s - pole) + np.conjugate(zeros[i]) / (s - np.conjugate(pole))
+        return resp
+
+    def plot_fit_linear(self, i, j, freqs=None):
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: Done in 5981299be06c6f9ac70790e63e4699ba0363ada6. Of course, I also changed the output back to dB. It was in linear scale before, which will be implemented in another function.

File: skrf/vectorFitting.py (hunk 1 above)
Comment (Vinc0110) [reply skrf/vectorFitting.py]: `subckt` is a SPICE convention, but it does not hurt to make it more readable by naming it write_spice_subcircuit_s. Done in 5981299be06c6f9ac70790e63e4699ba0363ada6

File: skrf/vectorFitting.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: I have changed the docstrings to Numpy notation in 9d3930697439866b098dd175343a58581c2077f4

**Comment by Vinc0110** [comment]
Does anyone know what's going on with those Python2.7 tests failing because of errors in all of the testcases in `test_network.NetworkTestCase`? This does not seem to be related to my contributions in this PR.

Error messages:

`TclError: no display name and no $DISPLAY environment variable`

Apparently, it has something to do with the Pylab configuration:

`nose.plugins.cover: ERROR: Coverage not available: unable to import coverage module
/home/runner/work/scikit-rf/scikit-rf/skrf/plotting.py:64: UserWarning: 
This call to matplotlib.use() has no effect because the backend has already
been chosen; matplotlib.use() must be called *before* pylab, matplotlib.pyplot,
or matplotlib.backends is imported for the first time.`

Is this a common problem with Python2.7? It seems to be fine with Python3.x

File: skrf/vectorFitting.py (hunk 2 above)
Comment (Vinc0110) [reply skrf/vectorFitting.py:28]: I don't think that's a good idea. The model parameters are strongly related to a specific set of frequency responses, i.e. a `Network`, and most class functions use both the model parameters and the `Network` attributes, such as the frequency vector `Network.f` and the number of ports `Network.n_ports`. In fact, it is already pretty risky to allow the import of model parameters via `read_npz()`. Maybe it is even better to remove that function altogether and only allow the export via `write_npz()`?

**Review by jhillairet — APPROVED** [review]
Great notebooks and thanks for the docstring update ! To me it's OK. Maybe @arsenovic will comment further.

File: skrf/tests/test_vectorfitting.py [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+                                   [7.93470695e+10-4.54467471e+09j],
+                                   [8.19724835e+10-2.11876421e+10j]])
+        expected_props = np.array([-2.06451610e-15,
+                                   -2.45016478e-14,
+                                   -2.45016478e-14,
+                                   7.79744644e-13])
+        expected_const = np.array([-0.9871906,
+                                   -0.06043898,
+                                   -0.06043898,
+                                   -0.99401152])
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/tests/test_vectorfitting.py:22]: You're right, it's not needed anymore. Remove in 568942d6612de17889866e66b353c150449d95eb.

File: skrf/vectorFitting.py [hunk 10]
```diff
@@ -0,0 +1,856 @@
...
+
+        init_pole_spacing : str, optional
+            Type of initial pole spacing across the frequency interval of the S-matrix. Either linear (lin) or
+            logarithmic (log).
+
+        parameter_type : str, optional
+            Representation type of the frequency responses to be fitted. Either **scattering (S)**, **impedance (Z)**
+            or **admittance (Y)**. As scikit-rf can currently only read S parameters from a Touchstone file, the fit
+            should also be performed on the original S parameters. Otherwise, scikit-rf will convert the responses from
+            S to Z or Y, which might work for the fit but can cause other issues.
```
[the remark this answers is not in the document]
Comment (Vinc0110) [reply skrf/vectorFitting.py]: Good idea, thanks! Implemented in c50c6c9a51265581819537a00f864c654c7f6f01.

**Comment by jhillairet** [comment]
Thanks for the latest additions and updates of the documentation. Did you check that all the equations render correctly in HTML? (sometimes there are bugs which require putting small tuning such as extra blank lines, etc)

Do you have ideas why the Python2 test fails? 

If not, as the Python2 version will be progressively not supported, I think we could skip this test for Python2, using `unittest.skipIf` decorator in front of your test and importing `six`:

```
import six

class VectorFittingTestCase(unittest.TestCase):

    @unittest.skipIf(six.PY2, "Not Compatible with Python 2")
    def test_vectorfitting_ring_slot(self):
```

**Comment by jhillairet** [comment]
in fact the Python2 errors are not related to your code, sorry. It passes with Travis but not with Github Action, so it seems more related to github Action configuration. So forget my previous comment.

Do you think it's OK to merge?

**Comment by Vinc0110** [comment]
The HTML should be OK, at least it is fine when I compile the docs locally.
You go go ahead and merge the PR.

**Comment by jhillairet** [comment]
> in fact the Python2 errors are not related to your code, sorry. It passes with Travis but not with Github Action, so it seems more related to github Actions configuration. So forget my previous comment.

For info, this has been fixed in #417 .

**Comment by Vinc0110** [comment]
All checks have finally passed. 

@jhillairet The Python2.7 tests were still failing with your changes in https://github.com/scikit-rf/scikit-rf/pull/417. What did the trick was to load `skrf.plotting` before loading `matplotlib.pyplot`. It seems like `plotting.py` is changing the matplotlib backend to `Agg` is no display is found. This has to happen first

**Comment by jhillairet** [comment]
> @jhillairet The Python2.7 tests were still failing with your changes in #417. What did the trick was to load `skrf.plotting` before loading `matplotlib.pyplot`. It seems like `plotting.py` is changing the matplotlib backend to `Agg` if no display is found. This has to happen first.

Oh, great, good catch indeed !

**Comment by jhillairet** [comment]
Great addition, thank you!

**Comment by Vinc0110** [comment]
What is the procedure now? I'm curious to see this feature in the actual skrf Conda package and on readthedocs. When/How is that going to happen?

**Comment by jhillairet** [comment]
> What is the procedure now? I'm curious to see this feature in the actual skrf Conda package and on readthedocs. When/How is that going to happen?

Well, I'm making new releases on Pip and Conda on a non-regularly basis, when few modifications have been made. I think this addition deserves indeed a new release. I'll try to do it this week-end.

**Comment by jhillairet** [comment]
scikit-rf 0.16.0 had been released on pip and conda (forge) and includes your PR.

**Comment by Vinc0110** [comment]
Awesome! Thank you for your help.

**Pull request merged**